pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,194 @@
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(*labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(*labels)
+	}
+	newlabels, err := c.generateLabels(ctx, pod)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge, err := podGaugeVec.GetMetricWith(*newlabels)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new gauge: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge.Set(float64(1))
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("podmetrics").
+		For(&v1.Pod{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10000}).
+		Complete(c)
+	return err
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(ctx context.Context, pod *v1.Pod) (*prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[podName] = pod.GetName()
+	metricLabels[podNameSpace] = pod.GetNamespace()
+	// Selflink has been deprecated after v.1.20
+	// Manually generate the selflink for the first owner reference
+	// Currently we do not support multiple over references
+	selflink := ""
+	if len(pod.GetOwnerReferences()) != 0 {
+		ownerreference := pod.GetOwnerReferences()[0]
+		selflink = fmt.Sprintf("/apis/%s/namespaces/%s/%ss/%s", ownerreference.APIVersion, pod.Namespace, strings.ToLower(ownerreference.Kind), ownerreference.Name)
+	}
+	metricLabels[ownerSelfLink] = selflink
+	metricLabels[podHostName] = pod.Spec.NodeName
+	metricLabels[podPhase] = string(pod.Status.Phase)
+	provisioner := v1alpha5.DefaultProvisioner
+	if name, ok := pod.Spec.NodeSelector[v1alpha5.ProvisionerNameLabelKey]; ok {
+		provisioner.Name = name
+	}
+	metricLabels[podProvisioner] = provisioner.Name
+	nodename := types.NamespacedName{Name: pod.Spec.NodeName}
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, nodename, node); err != nil {
+		metricLabels[podHostZone] = "N/A"
+		metricLabels[podHostArchitecture] = "N/A"
+		metricLabels[podHostCapacityType] = "N/A"
+		metricLabels[podHostInstanceType] = "N/A"
+	} else {
+		metricLabels[podHostZone] = node.Labels[v1.LabelTopologyZone]
+		metricLabels[podHostArchitecture] = node.Labels[v1.LabelArchStable]
+		metricLabels[podHostCapacityType] = node.Labels[v1alpha5.LabelCapacityType]
+		metricLabels[podHostInstanceType] = node.Labels[v1.LabelInstanceTypeStable]
+	}
+	// Add pod labels
+	labels, err := json.Marshal(pod.GetLabels())
```

Comment:
Why do we want to jsonify this? Does this always result in the same string (i.e. map keys are not ordered).

Reply:
We discussed  and want to also generate metrics according to pod's k8s labels. We talked about adding them to the prometheus labels dynamically. However, after doing some research I am convinced that it is an extrem anti-pattern  to add labels dynamically.

The solution is to put the k8s labels as the value of one prometheus label (i.e., podLabels). This allow us to do metric filtering using the podLabels prometheus label. Therefore, we need to convert he k8s labels into one single string. Json is just one way to do it. we can also just do simple key value pairs. But this works pretty well.

BTW, we don't really care about the order of these k8s lables.